module/zfs/dsl_scan.c

```diff
@@ -344,6 +346,112 @@ bp2sio(const blkptr_t *bp, scan_io_t *sio, int dva_i)
+	uint64_t asize = sio->sio_asize;
+
+	ASSERT(MUTEX_HELD(&queue->q_vd->vdev_scan_io_queue_lock));
+
+	if ((found_sio = avl_find(&queue->q_zios_by_addr, sio, &idx)) != NULL) {
+		/* block is already scheduled for reading */
+		if (sio->sio_dummy)
+			found_sio->sio_dummy = B_TRUE;
+		mutex_enter(&queue->q_scn->scn_status_lock);
+		queue->q_scn->scn_bytes_pending -= sio->sio_asize;
+		mutex_exit(&queue->q_scn->scn_status_lock);
+		kmem_free(sio, sizeof (*sio));
+		return;
+	}
+	avl_insert(&queue->q_zios_by_addr, sio, idx);
+	atomic_add_64(&queue->q_zio_bytes, asize);
+
+	range_tree_set_gap(queue->q_exts_by_addr, zfs_scan_max_ext_gap);
+	range_tree_add_fill(queue->q_exts_by_addr, offset, asize, asize);
+}
+
+/*
+ * Given all the info we got from our metadata scanning process, we
+ * construct a scan_io_t and insert it into the scan sorting queue. The
+ * I/O must already be suitable for us to process. This is controlled
+ * by dsl_scan_enqueue().
+ */
+static void
+scan_io_queue_insert(dsl_scan_io_queue_t *queue, const blkptr_t *bp, int dva_i,
+    int zio_flags, const zbookmark_phys_t *zb, boolean_t dummy)
+{
+	scan_io_t *sio = kmem_zalloc(sizeof (*sio), KM_SLEEP);
+
+	ASSERT(MUTEX_HELD(&queue->q_vd->vdev_scan_io_queue_lock));
+
+	bp2sio(bp, sio, dva_i);
+	sio->sio_flags = zio_flags;
+	sio->sio_zb = *zb;
+	sio->sio_dummy = dummy;
+
+	/*
+	 * Increment the bytes pending counter now so that we can't
+	 * get an integer underflow in case the worker processes the
+	 * zio before we get to incrementing this counter.
+	 */
+	mutex_enter(&queue->q_scn->scn_status_lock);
+	queue->q_scn->scn_bytes_pending += sio->sio_asize;
+	mutex_exit(&queue->q_scn->scn_status_lock);
+
+	scan_io_queue_insert_impl(queue, sio);
+}
+
+static void
+dsl_scan_enqueue_impl(dsl_pool_t *dp, const blkptr_t *bp, int zio_flags,
+    const zbookmark_phys_t *zb, boolean_t dummy)
+{
+	int i;
+	spa_t *spa = dp->dp_spa;
+
+	ASSERT(dp->dp_scan != NULL);
+	ASSERT(dp->dp_scan->scn_is_sorted);
+
+	for (i = 0; i < BP_GET_NDVAS(bp); i++) {
+		dva_t dva;
+		vdev_t *vdev;
+
+		dva = bp->blk_dva[i];
+		vdev = vdev_lookup_top(spa, DVA_GET_VDEV(&dva));
+		ASSERT(vdev != NULL);
+
+		mutex_enter(&vdev->vdev_scan_io_queue_lock);
+		if (vdev->vdev_scan_io_queue == NULL)
+			vdev->vdev_scan_io_queue = scan_io_queue_create(vdev);
+		scan_io_queue_insert(vdev->vdev_scan_io_queue, bp,
+		    i, zio_flags, zb, dummy);
+		mutex_exit(&vdev->vdev_scan_io_queue_lock);
+	}
+}
+
+/*
+ * Given a set of I/O parameters as discovered by the metadata traversal
+ * process, attempts to place the I/O into the sorted queues (if allowed),
+ * or immediately executes the I/O. The dummy flag can be set to
+ */
+static void
+dsl_scan_enqueue(dsl_pool_t *dp, const blkptr_t *bp, int zio_flags,
+    const zbookmark_phys_t *zb, boolean_t dummy)
+{
+	ASSERT(!BP_IS_EMBEDDED(bp));
+	if (!dp->dp_scan->scn_is_sorted) {
+		if (dummy)
+			return;
+
+		scan_exec_io(dp, bp, zio_flags, zb, NULL);
+		return;
+	}
```

Comment:
possibly "else" rather than "return", just to make the structure a little more obvious?